addOns/websocket/src/main/java/org/zaproxy/zap/extension/websocket/pscan/WebSocketPassiveScanThread.java

```diff
@@ -0,0 +1,189 @@
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.db.DatabaseException;
+import org.zaproxy.zap.extension.websocket.alerts.WebSocketAlertWrapper;
+import org.zaproxy.zap.extension.websocket.WebSocketMessage;
+import org.zaproxy.zap.extension.websocket.WebSocketMessageDTO;
+import org.zaproxy.zap.extension.websocket.WebSocketProxy;
+import org.zaproxy.zap.extension.websocket.WebSocketSenderListener;
+import org.zaproxy.zap.extension.websocket.db.TableWebSocket;
+
+import java.util.Iterator;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Implements a background thread for passive scanning
+ */
+public class WebSocketPassiveScanThread extends Thread implements WebSocketSenderListener {
+    
+    private static final Logger LOGGER = Logger.getLogger(WebSocketPassiveScanThread.class);
+    
+    public static final int WEBSOCKET_LISTENER_ORDER = 20; //TODO Make it more formal
+    
+    /**
+     * Interval for message reading
+     */
+    public static final int SLEEP_TIME = 5000;
+    
+    /**
+     * Thread safe Linked Queue
+     */
+    private LinkedBlockingQueue<MessageWrapper> messagesBuffer;
+    
+    /**
+     * {@code True} to enable the passive scan Thread
+     */
+    private volatile boolean isPassiveScannerActive;
+    
+    /**
+     * Reference to Database. Used in order to pick messages
+     * for scanning
+     */
+    private TableWebSocket tableWebSocket;
+    
+    /**
+     * Manager used to updating the messages table. In addition,
+     * used by passive scan thread so as to get access at plugins
+     */
+    private WebSocketPassiveScannerManager passiveScannerManager;
+    
+    /**
+     * Initialize the passive scan in background thread. By default thread is
+     * inactive and not alive. In order to activate thread set
+     * {@link WebSocketPassiveScanThread#setPassiveScannerActive(boolean)} true and
+     * start thread with {@link WebSocketPassiveScanThread#start()}.
+     * @param passiveScannerManager the manager
+     */
+    public WebSocketPassiveScanThread (WebSocketPassiveScannerManager passiveScannerManager) {
+        this.passiveScannerManager = passiveScannerManager;
+        this.isPassiveScannerActive = false;
+        messagesBuffer = new LinkedBlockingQueue<>();
+    }
+    
+    /**
+     * @return true if the table was initialized
+     */
+    public boolean hasTableWebSocket(){
+        return tableWebSocket != null;
+    }
+    
+    public void setTableWebSocket(TableWebSocket tableWebSocket) {
+        this.tableWebSocket = tableWebSocket;
+    }
+    
+    public TableWebSocket getTableWebSocket() {
+        return tableWebSocket;
+    }
+    
+    @Override
+    public int getListenerOrder() {
+        return WEBSOCKET_LISTENER_ORDER;
+    }
+    
+    @Override
+    public void onMessageFrame(int channelId, WebSocketMessage message, WebSocketProxy.Initiator initiator) {
+        //TODO: Take care {@link WebSocketMessage#OPCODE_CONTINUATION}
```

Comment:
@ManosMagnus is there still something outstanding here?

Reply:
`OPCODE_CONTINUATION` denotes this message is a continuation fragment of the previous message. That means that the current message is part of a bigger one. That either can be ignored or can be handled with a buffer list. In passive scanner case I can not imagine any case in which is really useful to be handled by. I added the TODO so as if any issue were (ever) arisen related with, we could easily come up with the answer. If you think that is essential to be handled let me know. @thc202 can help us with that also.

Reply:
We can check if the message is finished before adding it to the queue (that would address the TODO).

Reply:
Ok that will be great. But how we can know that the message is incomplete? Only if we keep the message in a "pending" mode in order to check the next message's opcode. Am I missing something?

Reply:
My fault, I can use WebSocketMessage.isFinished()